### DISCUSSION TITLE
New snippet syntax

### DISCUSSION THREAD

**Discussion by orbitalquark** [orbitalquark/textadept#407] [General]
The next nightly build of Textadept will support TextMate- and LSP-style snippets (`${1:default}`, `${2/.+/${0:/upcase}/}`, etc.) as described [here][1]. The legacy snippet format (`%1(default)`, `%2<lua transform code>`, etc.) is deprecated, but still supported. There is a [migration guide][2] for migrating legacy snippets. Legacy support will be sometime removed after Textadept 12.0 is released, so please consider migrating earlier rather than later.

As a side effect, the [LSP][3] module now supports snippet completions.

[1]: https://orbitalquark.github.io/textadept/api.html#textadept.snippets
[2]: https://orbitalquark.github.io/textadept/api.html#migrating-legacy-snippets
[3]: https://github.com/orbitalquark/textadept-lsp

**Comment by orbitalquark** [comment]
The nightly build has failed, so this feature is not yet available. I hope to have it up later today. I'll update this discussion when it is.

**Reply by orbitalquark** [reply]
The new nightly builds are up now.

**Comment by paaguti** [comment]
Could you consider a simpler way of binding Lua snippets. I mean, currently 
I have a very simple and useful snippet function to insert the date, 
which I use as
```
    <date %<snip.today()>/>
```
in this line, I have other functions in my `snip` module. There are systems,
where relying on an external Lua interpreter may be difficult (due to 
'alleged security' policies). Besides, Textadept aleady has a Lua engine.
Why not use it?

Just another example, which might be more difficult...

```
%<snip.ifempty(buffer.filename,"%% ",text)>\bibliography{%<snip.file_no_ext(snip.basename(buffer.filename))>}
```

IMHO, we loose this extremely nice integration of snippets with the "rest of" textadept

**Reply by paaguti** [reply]
A possibility would be to have 
```
$`lua code inside here`
```
The first snippet above would look like

```
$`snip.today()`
```

**Comment by orbitalquark** [comment]
Yes, that is a good idea. I was thinking to experiment with a '```' 
syntax extension for Lua code instead of Shell code, but forgot about 
it. Thanks for the reminder.

**Comment by orbitalquark** [comment]
Lua code is supported in snippets using ```Lua code``` syntax: 
https://github.com/orbitalquark/textadept/commit/846de7b1b648ef7e1266d1b76b41b358d8065026

**Comment by paaguti** [comment]
Good to hear :-) I'll test and report back. I was thinking of

$`lua_code()`

because it would look more uniform. I'll try out

Thanks, /PA

On Mon, 12 Jun 2023 at 18:08, orbitalquark ***@***.***> wrote:

> Lua code is supported in snippets using ```Lua code``` syntax:
>
> https://github.com/orbitalquark/textadept/commit/846de7b1b648ef7e1266d1b76b41b358d8065026
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/407#discussioncomment-6154289>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4G3TFSNHMP44C7Z53LXK45JJANCNFSM6AAAAAAZBXT7KA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Fragen sind nicht da, um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

Headaches with a Juju log:
unit-basic-16: 09:17:36 WARNING juju.worker.uniter.operation we should run
a leader-deposed hook here, but we can't yet

**Comment by paaguti** [comment]
I tried out and can do most. There is however a case which I can't cover with transforms or the current Lua proposal.
I have a snippet like:

```
%5<snip.ifempty(text,"% ")>\title{%5}
```
which expands to "insert '% ' if the capture 5 is empty, else insert an empty string". That would be

```
% \title{}
```

or 

```
\title{My paper}
```

I can't find an easy way to implement this in the new syntax. Yasnippets in emacs  would do this as

```
${5:$(yas-if-empty "%")}\title{$5}
```

because they pass the text of the caption internally. But something like that or 

```
${5:```snip.ifempty(text,"% ")```}\title{$5}
```

would be helpful

**Reply by orbitalquark** [reply]
With https://github.com/orbitalquark/textadept/commit/d3e5955b5e4b09deb2f955a2223ab45c6553114a, you can do this:

`${1/.*/${0:?:% }/}\\title{$1}`

With the current nightly you would do this:

`${1/.*/${0:? :% }/}\\title{$1}` (note the space after `?`)

From the documentation:

A “${_n_:?_if_:_else_}” sequence [...] inserts _if_ if the content of capture _n_ is non-empty. Otherwise, _else_ is used. In your case, _if_ should be empty if $1 is present, and "% " if $1 is not.

**Reply by paaguti** [reply]
It would be wise to expand on the `: ` (space behind). That was my breaking point.  I still prefer my `snip.if_empty()` solution, because it is written and reads easier, but at least the new notation does what it promises ;-)

**Reply by paaguti** [reply]
I know I'm a pain... but porting my snippets is being an idem ;-)

I have this:
```
%6<snip.ifdefault(text, "volume", "OPT")>volume = {%6(volume)},
```
which will insert `OPT` at the beginning of the line if I don't touch the default value. If I had something like:
```
${6:`snip.ifdefault(text, "volume", "OPT")`}volume = {${6:volume}}
```
or (more emacs-ish):
```
${6:$(snip.ifdefault(text, "volume", "OPT"))}volume = {${6:volume}}
```
I would be able to keep a 'couple' of years worth of work ;-)

Of course, access to the capture in Lua like `text` would be most welcome

**Reply by orbitalquark** [reply]
You are not being a pain at all. It helps verify the new system is working properly.

What about this:

```
${6/volume/OPT/}volume = {${6:volume}},
```

**Reply by paaguti** [reply]
It doesn't do what I really need... Just checked and the output is
```
  vovolume = {vo},
```

What I'd need would be
```
  volume = {vo},
```
i.e. nothing when I write something different from 'volume' in the capture. The other thing that doesn't fit is
```
  OPT2volume = {volume2},
```
i.e. when 'volume' is part of the capture.

My `snip.ifdefault` function returns the 3rd argument  if the capture (1st argument, text) is equal to the 2nd argument, else it returns an empty string. I have a full collection of bib(la)tex snippets that use it...

**Reply by orbitalquark** [reply]
You need to be more specific with examples then. You said "will insert OPT at the beginning of the line if I don't touch the default value". In the snippet I gave, you're not touching the default value of the $6 placeholder, and it's inserting "OPT" at the beginning of the line (a seemingly satisfactory result). The examples you're giving appear to contradict your stated expectation. At least, that's just how I'm reading it.

**Comment by paaguti** [comment]
Right, apologies, but wrapping up before the summer is being sort of a strain... 


Anyhow, let's see if I can make my case for Lua code referencing captures in snippets:

Imagine I have my `snip` module and, inside it, I have this function:
```
function M.ifdefault(txt, defv, val)
   if txt and string.len(txt) > 0 and txt ~= defv then
      return ""
   else
      return val1
   end
end
```

Then, my proposed snippet syntax:
```
${6:`snip.ifdefault(text, "volume", "OPT")`}volume = {${6:volume}}
```
would allow to insert `OPT` before `volume` at the beginning of the line, if (and only if) the field $6 contains the default value (i.e. `volume`). 

I need this kind of programmatic text transforms and I really don't know how to write them with the new syntax. With the years of using Textadept, I have a collected a lot of these things that make my life really easy. It would be (a bit more than just) a pity to loose them.

**Reply by rgieseke** [reply]
I think it's always worth to think about the best API design, but aside of that, would it be possible for you to keep the old snippet code from Textadept as a compatibility model along with your snippets?

**Comment by orbitalquark** [comment]
> Imagine I have my `snip` module and, inside it, I have this function:
> 
> ```
> function M.ifdefault(txt, defv, val)
>    if txt and string.len(txt) > 0 and txt ~= defv then
>       return ""
>    else
>       return val1
>    end
> end
> ```
> 
> Then, my proposed snippet syntax:
> 
> ```
> ${6:`snip.ifdefault(text, "volume", "OPT")`}volume = {${6:volume}}
> ```
> 
> would allow to insert `OPT` before `volume` at the beginning of the line, if (and only if) the field $6 contains the default value (i.e. `volume`).

I believe this might work:

```
${6/(volume)|.*/${1:?:OPT}/}volume = ${6:volume}
```

The $6 transform tries to match "volume" (the default) and captures the result to register 1. If that fails ("volume" does not match), the transform matches everything (".*") and stores that to register 0. In the replacement, the transform looks in register 1 for a non-empty value ("volume"). If that value exists, nothing is given as a replacement. Otherwise, replace with "OPT". The contents of register 0 (any typed text) is discarded.

This should satisfy the "if and only if" criteria.

> I need this kind of programmatic text transforms and I really don't know how to write them with the new syntax. With the years of using Textadept, I have a collected a lot of these things that make my life really easy. It would be (a bit more than just) a pity to loose them.

My goal is to help find a 1-to-1 translation of existing legacy snippets to the new format. I wouldn't want you to lose them either.

**Reply by paaguti** [reply]
> My goal is to help find a 1-to-1 translation of existing legacy snippets to the new format. I wouldn't want you to lose them either.

I know, and that's why I'm bugging you with this Lua stuff. I love the regex syntax and I will use it, but it is tricky. As for transforms, they are lovely, but limited (i.e. no extra params). But some text transformations are easier, quicker and cleaner to write in a programming language (Lua in the case of Textadept, LISP for emacs, you name it).  Some may even prove impossible to write with a regex or might yield an expression which you may need to keep documented somewhere, in order not to forget.
Snippets should be there to speed up writing, but what is the gain if I need to sit an afternoon to design, test and debug a regex in snippet? ;-)

Thx again, /PA

PS: the last proposal doesn't work either :(
```
  OPTnumber = {4},
  pages = {pages},
  2month = {month2},
```
It seems to do sort of the opposite of what I need

**Reply by orbitalquark** [reply]
Okay, sorry about that. I must have read it wrong. Try this:

```
${6/^(volume)$|.*/${1:+OPT}/}volume = ${6:volume}
```

`^(volume)$` requires an exact match. The `${1:+OPT}` replacement uses "OPT" *only if* capture 1 exists ("volume").

You could also have `${1:?OPT:}` (switch the "if" and "else" cases in my previous example since I did the opposite of what you expected), but it simplifies to the `+` operator I gave.

With the above, I get:

```
volume = {4}
volume = {pages}
volume = {month2}
OPTvolume = {volume}
```

**Reply by orbitalquark** [reply]
I agree that some snippets may be less readable without a higher level language like Lua. I may look into adding transform-time Lua code execution, but it is not on my priority list.

**Reply by paaguti** [reply]
Thanks a lot! That worked. 
So that might be a snippet for snippets ;-)
Anyhow, do not complete forget transform-time Lua completely ;-) It is the split between convenient and outstanding.
... at least for me (TM)
